From 44d5c13515c04d134670ff9b3b02cc478bd90aa5 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Tue, 13 Jul 2021 10:56:56 +0800
Subject: [PATCH 70/98] drm-backend: Add sw-cursor config to force software
 cursor

Add configuration option to force software cursor rendering in DRM
backend, useful for hardware with broken cursor planes or debugging.

Tested with:
[core]
sw-cursor=true

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 frontend/main.c                 | 2 ++
 include/libweston/backend-drm.h | 2 ++
 libweston/backend-drm/drm.c     | 2 ++
 3 files changed, 6 insertions(+)

diff --git a/frontend/main.c b/frontend/main.c
index 7f8f97d0..2e409e3c 100644
--- a/frontend/main.c
+++ b/frontend/main.c
@@ -4208,6 +4208,8 @@ load_drm_backend(struct weston_compositor *c, int *argc, char **argv,
 	                               &config.pageflip_timeout, 0);
 	weston_config_section_get_bool(section, "pixman-shadow",
 				       &config.use_pixman_shadow, true);
+	weston_config_section_get_bool(section, "sw-cursor",
+				       &config.use_sw_cursor, false);
 	if (without_input)
 		c->require_input = !without_input;
 
diff --git a/include/libweston/backend-drm.h b/include/libweston/backend-drm.h
index 62cd3e69..400a10d8 100644
--- a/include/libweston/backend-drm.h
+++ b/include/libweston/backend-drm.h
@@ -266,6 +266,8 @@ struct weston_drm_backend_config {
 	 * "color-management" to be enabled.
 	 */
 	bool offload_blend_to_output;
+
+	bool use_sw_cursor;
 };
 
 #ifdef  __cplusplus
diff --git a/libweston/backend-drm/drm.c b/libweston/backend-drm/drm.c
index 5ff7696a..2213e101 100644
--- a/libweston/backend-drm/drm.c
+++ b/libweston/backend-drm/drm.c
@@ -4780,6 +4780,8 @@ drm_backend_create(struct weston_compositor *compositor,
 		goto err_drm_device;
 	}
 
+	device->cursors_are_broken |= config->use_sw_cursor;
+
 	/* A this point we have some idea of whether or not we have a working
 	 * cursor plane. */
 	if (!device->cursors_are_broken)
-- 
2.39.5

